### PR TITLE
8320890: [AIX] Find a better way to mimic dl handle equality

### PR DESCRIPTION
**Pull request by JoKern65**
Hi all,

This pull request contains a backport of commit [b8ae4a8c](https://github.com/openjdk/jdk/commit/b8ae4a8c0985d1763ac48ba78943d8b992d7be77) from the [openjdk/jdk](https://git.openjdk.org/jdk) repository.

The commit being backported was authored by Joachim Kern on 11 Jan 2024 and was reviewed by Thomas Stuefe and Martin Doerr.

Thanks!

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] [JDK-8320890](https://bugs.openjdk.org/browse/JDK-8320890) needs maintainer approval

### Issue
 * [JDK-8320890](https://bugs.openjdk.org/browse/JDK-8320890): [AIX] Find a better way to mimic dl handle equality (**Enhancement** - P4 - Approved)

### Reviewers
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk21u-dev.git pull/252/head:pull/252` \
`$ git checkout pull/252`

Update a local copy of the PR: \
`$ git checkout pull/252` \
`$ git pull https://git.openjdk.org/jdk21u-dev.git pull/252/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 252`

View PR using the GUI difftool: \
`$ git pr show -t 252`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk21u-dev/pull/252.diff">https://git.openjdk.org/jdk21u-dev/pull/252.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk21u-dev/pull/252#issuecomment-1936254576)

### CODE REVIEW

File: src/hotspot/os/aix/os_aix.cpp [hunk 1]
```diff
@@ -1126,7 +1128,8 @@ void *os::dll_load(const char *filename, char *ebuf, int ebuflen) {
     dflags |= RTLD_MEMBER;
   }
 
-  void * result= ::dlopen(filename, dflags);
+  const char* error_report = nullptr;
+  void * result = Aix_dlopen(filename, dflags, &error_report);
```
Comment (TheRealMDoerr) [inline src/hotspot/os/aix/os_aix.cpp:1132]: Diff to upstream version should get minimized: `void* result;` is in a separate line here: https://github.com/openjdk/jdk/commit/b8ae4a8c0985d1763ac48ba78943d8b992d7be77

File: src/hotspot/os/aix/porting_aix.cpp [hunk 2]
```diff
@@ -891,3 +897,274 @@ bool AixMisc::query_stack_bounds_for_current_thread(stackbounds_t* out) {
...
+      // therefore nag
+      assert(false, "os::pd_dll_unload() library was not loaded by os::dll_load()");
+    }
+  }
+
+  // Update the dll cache
+  LoadedLibraries::reload();
+
+  return res;
+} // end: os::pd_dll_unload()
```
Comment (TheRealMDoerr) [inline src/hotspot/os/aix/porting_aix.cpp:1170]: Newline diff.

File: src/hotspot/os/posix/os_posix.cpp [hunk 3]
```diff
@@ -734,25 +735,19 @@ void os::dll_unload(void *lib) {
   if (l_path == nullptr) {
     l_path = "<not available>";
   }
-  int res = ::dlclose(lib);
+  char ebuf[1024];
+  bool res = os::pd_dll_unload(lib, ebuf, sizeof(ebuf));
```
Comment (TheRealMDoerr) [inline src/hotspot/os/posix/os_posix.cpp:740]: Empty lines are different than in upstream version.

**Review by TheRealMDoerr — COMMENTED** [review]
Looks correct, but please minimize diffs to upstream version!

File: src/hotspot/os/aix/porting_aix.cpp (hunk 2 above)
Comment (JoKern65) [reply src/hotspot/os/aix/porting_aix.cpp:1170]: This newline at the end of the file was marked red in my diff check, therefore I removed it. But OK I put it in again.

**Review by TheRealMDoerr — APPROVED** [review]
LGTM. Thanks!

**Comment by JoKern65** [comment]
/approval JDK-8320890 request I backport this to fix the bug on AIX. Not clean but trivial backport. Test runs on various platforms succeeded. Minor risk as the shared code changes are basically a refactoring, so the risk is limited to AIX.

**Pull request closed without merging**